distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+import asyncio
+import heapq
+import math
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Iterator
+from typing import Optional
+
+import dask
+
+from .scheduler import SchedulerPlugin, SchedulerState, TaskState, WorkerState
+
+
+class ActiveMemoryManager(SchedulerPlugin):
+    scheduler: SchedulerState
+    plugins: "set[ActiveMemoryManagerPlugin]"
+    interval: float
+
+    # Daemon that runs every <interval> seconds
+    run_task: asyncio.Task
+
+    # These attributes only exist within the scope of self.run()
+    # Current memory (in bytes) allocated on each worker, plus/minus pending actions
+    workers_memory: "dict[WorkerState, int]"
+    # Pending replications and deletions for each task
+    pending: "defaultdict[TaskState, tuple[set[WorkerState], set[WorkerState]]]"
+
+    def __init__(
+        self,
+        plugins: "Iterable[ActiveMemoryManagerPlugin]" = None,
+        interval: float = 2.0,
+    ):
+        self.interval = interval
+        self.plugins = set()
+        if plugins is None:
+            plugins = {ReplicateReduce(), Rebalance()}
+        for plugin in plugins:
+            self.register_plugin(plugin)
+
+    def register_plugin(self, plugin: "ActiveMemoryManagerPlugin") -> None:
+        plugin.manager = self
+        self.plugins.add(plugin)
```

Comment:
I would expect plugins might also finish in the future (like retiring workers).

Reply:
if by "finish" you mean unregister, I put in two examples of a plugin that detaches itself from the AMM when there's nothing left to do: ReplicateIncrease and WorkerShutdown